src/censored.jl

```diff
@@ -0,0 +1,519 @@
+#   ) / P_{x ~ d₀}(l ≤ x ≤ u),
+# where S[τ] is the entropy of τ.
+
+function entropy(d::LeftCensored)
+    d0 = d.uncensored
+    upper = d.upper
+    log_prob_upper = logccdf(d0, upper)
+    if value_support(typeof(d0)) === Discrete
+        logpu = logpdf(d0, upper)
+        log_prob_upper_inc = logaddexp(log_prob_upper, logpu)
+        xlogx_pu = xexpx(logpu)
+    else
+        log_prob_upper_inc = log_prob_upper
+        xlogx_pu = 0
+    end
+    entropy_bound = -xexpx(log_prob_upper_inc)
+    log_prob_trunc = log1mexp(log_prob_upper)
+    prob_trunc = exp(log_prob_trunc)
+    # truncation contains ~ no probability
+    prob_trunc < eps(one(log_prob_trunc)) && return entropy_bound
+    dtrunc = _to_truncated(d)
+    entropy_interval = prob_trunc * entropy(dtrunc) - xexpx(log_prob_trunc) + xlogx_pu
+    return oftype(entropy_bound, entropy_bound + entropy_interval)
+end
+function entropy(d::RightCensored)
+    d0 = d.uncensored
+    lower = d.lower
+    log_prob_lower_inc = logcdf(d0, lower)
+    if value_support(typeof(d0)) === Discrete
+        logpl = logpdf(d0, lower)
+        log_prob_lower = logsubexp(log_prob_lower_inc, logpl)
+        xlogx_pl = xexpx(logpl)
+    else
+        log_prob_lower = log_prob_lower_inc
+        xlogx_pl = 0
+    end
+    entropy_bound = -xexpx(log_prob_lower_inc)
+    log_prob_trunc = log1mexp(log_prob_lower)
+    prob_trunc = exp(log_prob_trunc)
+    # truncation contains ~ no probability
+    prob_trunc < eps(one(log_prob_trunc)) && return entropy_bound
+    dtrunc = _to_truncated(d)
+    entropy_interval = prob_trunc * entropy(dtrunc) - xexpx(log_prob_trunc) + xlogx_pl
+    return oftype(entropy_bound, entropy_bound + entropy_interval)
+end
+function entropy(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    log_prob_lower_inc = logcdf(d0, lower)
+    log_prob_upper = logccdf(d0, upper)
+    if value_support(typeof(d0)) === Discrete
+        logpl = logpdf(d0, lower)
+        logpu = logpdf(d0, upper)
+        log_prob_lower = logsubexp(log_prob_lower_inc, logpl)
+        log_prob_upper_inc = logaddexp(log_prob_upper, logpu)
+        xlogx_pl = xexpx(logpl)
+        xlogx_pu = xexpx(logpu)
+    else
+        log_prob_lower = log_prob_lower_inc
+        log_prob_upper_inc = log_prob_upper
+        xlogx_pl = xlogx_pu = 0
+    end
+    entropy_bound = -(xexpx(log_prob_lower_inc) + xexpx(log_prob_upper_inc))
+    log_prob_trunc = log1mexp(logaddexp(log_prob_lower, log_prob_upper))    
+    prob_trunc = exp(log_prob_trunc)
+    # truncation contains ~ no probability
+    prob_trunc < eps(one(log_prob_trunc)) && return entropy_bound
+    dtrunc = _to_truncated(d)
+    entropy_interval = prob_trunc * entropy(dtrunc) - xexpx(log_prob_trunc) + xlogx_pl + xlogx_pu
+    return entropy_bound + entropy_interval
+end
+
+
+#### Evaluation
+
+function pdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    px = float(pdf(d0, x))
```

Comment:
```suggestion
    px = pdf(d0, x)
```
should be fine, shouldn't it? This would allow to perform calculations with e.g. `Rational`s for discrete distributions.

Reply:
That being said, if each conditional calls `pdf` and `cdf` or `ccdf`, zeroing out the unneeded quantity, we get type-stability and end up with a lot of `Const(0.0)`s. Perhaps the compiler then can optimize away the calls where values are unused.

Reply:
e.g.
```julia
function pdf(d::Censored, x::Real)
    d0 = d.uncensored
    lower = d.lower
    upper = d.upper
    return if _in_open_interval(x, lower, upper)
        zero(cdf(d0, x)) + pdf(d0, x)
    elseif x == lower
        cdf(d0, x) + zero(pdf(d0, x))
    elseif x == upper
        px = pdf(d0, x)
        ccdf(d0, x) + (value_support(typeof(d0)) === Discrete ? px : zero(px))
    else  # not in support
        zero(cdf(d0, x)) + zero(pdf(d0, x))
    end
end
```

Reply:
Just checked for `censored(Normal(0, 1); -1, 1)`, and the version above is 15-60% slower than the one currently used in the code.